### PR TITLE
[WIP] XGBoost

### PR DESCRIPTION
**Pull request by TirelessClock**
XGBoost (eXtreme Gradient Boosting) is an optimized implementation of Gradient Boosting. 
Working forward from PR #3735 where I implement Gradient Boosting.

### CODE REVIEW

**Review by zoq — COMMENTED** [review]
The failing CI is because of the other PR that we have to merge first?

File: src/mlpack/tests/xgboost_test.cpp [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /**
  * @file tests/xgboost_test.cpp
- * @author Rishabh Garg
+ * @author Abhimanyu Dayal
```
Comment (zoq) [inline src/mlpack/tests/xgboost_test.cpp:3]: Since this is shared code, I would add you as an additional author, what do you think?

Comment (TirelessClock) [reply src/mlpack/tests/xgboost_test.cpp:3]: Yeah sure, this works!

File: src/mlpack/methods/xgboost/feature_importance.hpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+   
+  map<size_t, size_t> featureFrequency;
+  map<size_t, double> featureCover;
+  vector<size_t> rankByFrequency;
+  vector<size_t> rankByCover;
+
+  // Empty class constructor.
+  FeatureImportance() { /*Nothing to do*/ }
+
+  //! Edit the featureFrequency value.
```
[the remark this answers is not in the document]
Comment (TirelessClock) [reply src/mlpack/methods/xgboost/feature_importance.hpp:50]: I'm sorry, I don't understand. Could you please give me an example?

**Comment by TirelessClock** [comment]
> The failing CI is because of the other PR that we have to merge first?

@zoq It's failing "Documentation Build and Test", but I'm not sure how to fix that.

**Comment by zoq** [comment]
> > The failing CI is because of the other PR that we have to merge first?
> 
> @zoq It's failing "Documentation Build and Test", but I'm not sure how to fix that.

Looking at the output:

```
Processing	file:///home/jenkins/workspace/pull-request%20documentation%20build%20and%20test/doc/html/user/methods/xgboost.html


List of broken links and other issues:
file:///home/jenkins/workspace/pull-request%20documentation%20build%20and%20test/doc/html/user/methods/gradient_boosting.html	
  Line: 398
  Code: 404 File `/home/jenkins/workspace/pull-request documentation build and test/doc/html/user/methods/gradient_boosting.html' does not exist
 To do: The link is broken. Double-check that you have not made any typo,
	or mistake in copy-pasting. If the link points to a resource that
	no longer exists, you may want to remove or fix the link.

https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/grad_boosting/grad_boosting.hpp	
  Line: 647
  Code: 404 Not Found
 To do: The link is broken. Double-check that you have not made any typo,
	or mistake in copy-pasting. If the link points to a resource that
	no longer exists, you may want to remove or fix the link.
```


it tries to include a file from master that does not exsist.

**Comment by TirelessClock** [comment]
> > > The failing CI is because of the other PR that we have to merge first?
> > 
> > 
> > @zoq It's failing "Documentation Build and Test", but I'm not sure how to fix that.
> 
> Looking at the output:
> 
> ```
> Processing	file:///home/jenkins/workspace/pull-request%20documentation%20build%20and%20test/doc/html/user/methods/xgboost.html
> 
> 
> List of broken links and other issues:
> file:///home/jenkins/workspace/pull-request%20documentation%20build%20and%20test/doc/html/user/methods/gradient_boosting.html	
>   Line: 398
>   Code: 404 File `/home/jenkins/workspace/pull-request documentation build and test/doc/html/user/methods/gradient_boosting.html' does not exist
>  To do: The link is broken. Double-check that you have not made any typo,
> 	or mistake in copy-pasting. If the link points to a resource that
> 	no longer exists, you may want to remove or fix the link.
> 
> https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/grad_boosting/grad_boosting.hpp	
>   Line: 647
>   Code: 404 Not Found
>  To do: The link is broken. Double-check that you have not made any typo,
> 	or mistake in copy-pasting. If the link points to a resource that
> 	no longer exists, you may want to remove or fix the link.
> ```
> 
> it tries to include a file from master that does not exsist.

Oh ok. I'll get on this ASAP. I think I had the same issue on my other PR too.

File: src/mlpack/tests/xgboost_test.cpp (hunk 1 above)
Comment (TirelessClock) [reply src/mlpack/tests/xgboost_test.cpp:3]: Actually, his original work in this file was a redundant to my implementation, so there isn't any shared code in this file. 
He had originally written an SSE loss function, and test cases for that, which I'm not using in my XGBoost implementation.

**Comment by TirelessClock** [comment]
@zoq Hello, can this PR be reopened?
